internal/verifier/oplog_reader.go

```diff
@@ -0,0 +1,716 @@
+				if err != nil {
+					return errors.Wrap(err, "extracting o2._id from update")
+				}
+			} else {
+				return errors.Wrap(err, "extracting o._id from update")
+			}
+		default:
+			panic(fmt.Sprintf("op=%#q unexpected (%v)", opName, rawDoc))
+		}
+
+		if docField != "" {
+			if opName == "u" {
+				opName = "r"
+			}
+
+			doc, err := mbson.Lookup[bson.Raw](rawDoc, docField)
+			if err != nil {
+				return errors.Wrap(err, "extracting doc from op")
+			}
+
+			docLength = option.Some(types.ByteCount(len(doc)))
+			docID, err = doc.LookupErr("_id")
+			if err != nil {
+				return errors.Wrap(err, "extracting doc ID from op")
+			}
+		} else {
+			if docID.IsZero() {
+				panic("zero doc ID!")
+			}
+		}
+
+		docID.Value = slices.Clone(docID.Value)
+
+		events = append(
+			events,
+			ParsedEvent{
+				OpType:      oplogOpToOperationType[opName],
+				Ns:          NewNamespace(SplitNamespace(nsStr)),
+				DocID:       docID,
+				FullDocLen:  docLength,
+				ClusterTime: lo.ToPtr(ts),
+			},
+		)
+
+		return nil
+	}
+
+	for _, rawDoc := range o.curDocs {
+		opName, err := mbson.Lookup[string](rawDoc, "op")
+		if err != nil {
+			return nil, bson.Timestamp{}, err
+		}
+
+		err = mbson.LookupTo(rawDoc, &latestTS, "ts")
+		if err != nil {
+			return nil, bson.Timestamp{}, err
+		}
+
+		switch opName {
+		case "n":
+			// Ignore.
+		case "c":
+			oDoc, err := mbson.Lookup[bson.Raw](rawDoc, "o")
+			if err != nil {
+				return nil, bson.Timestamp{}, err
+			}
+
+			el, err := oDoc.IndexErr(0)
+			if err != nil {
+				return nil, bson.Timestamp{}, errors.Wrap(err, "getting first el of o doc")
+			}
+
+			cmdName, err := el.KeyErr()
+			if err != nil {
+				return nil, bson.Timestamp{}, errors.Wrap(err, "getting first field name of o doc")
+			}
+
+			if cmdName != "applyOps" {
+				if o.onDDLEvent == onDDLEventAllow {
+					o.logIgnoredDDL(rawDoc)
+					continue
+				}
+
+				if !latestTS.After(allowDDLBeforeTS) {
+					o.logger.Info().
+						Stringer("event", rawDoc).
+						Msg("Ignoring unrecognized write from the past.")
+
+					continue
+				}
+
+				return nil, bson.Timestamp{}, UnknownEventError{rawDoc}
+			}
+
+			var opsArray bson.Raw
+			err = mbson.UnmarshalElementValue(el, &opsArray)
+			if err != nil {
+				return nil, bson.Timestamp{}, errors.Wrap(err, "parsing applyOps")
+			}
+
+			arrayVals, err := opsArray.Values()
+			if err != nil {
+				return nil, bson.Timestamp{}, errors.Wrap(err, "getting applyOps values")
+			}
+
+			// Might as well ...
+			events = slices.Grow(events, len(arrayVals))
+
+			for i, opRV := range arrayVals {
+				opRaw, err := mbson.CastRawValue[bson.Raw](opRV)
+				if err != nil {
+					return nil, bson.Timestamp{}, errors.Wrapf(err, "extracting applyOps[%d]", i)
+				}
+
+				opName, err := mbson.Lookup[string](opRaw, "op")
+				if err != nil {
+					return nil, bson.Timestamp{}, errors.Wrapf(err, "extracting applyOps[%d].op", i)
+				}
+
+				err = parseOneDocumentOp(opName, latestTS, opRaw)
```

Comment:
[note] I imagine because this logic is only used by the verifier it is ok for rechecks to be spuriously performed on documents which are attempted to be modified by a multi-statement transaction which ultimately aborts. As a comparison point, change streams only emits change event for multi-statement transactions after they are committed.

This behavior of the oplog reader is also what enables its resume token to only be aware of the 'ts' timestamp position within the oplog collection.

I'm unsure if this is the best spot to call this out or if there's some high-level overview on how the oplog reader mode behaves it can be summarized in instead.

Reply:
I added a note at the top of this file about the 2 cases (including this) where oplog reader triggers “extra” rechecks.